// winccoa-manager/lib/binding.js
"use strict";
// disable rules that trigger because the AddOn does not have type definitions
/* eslint-disable @typescript-eslint/no-unsafe-return,
                  @typescript-eslint/no-unsafe-assignment,
                  @typescript-eslint/no-unsafe-call,
                  @typescript-eslint/no-unsafe-return,
                  @typescript-eslint/no-unsafe-member-access
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WinccoaManager = exports.WinccoaElementType = exports.WinccoaSecurityEventId = exports.WinccoaConnectUpdateType = void 0;
//@ts-expect-error Importing an AddOn without type declaratiob
const winccoaconnection_1 = require("C:/Siemens/Automation/WinCC_OA/3.20/bin/winccoaconnection");
const error_1 = require("./error");
const connection_binding_1 = require("./connection-binding");
/**
 * Enumerates the type of updates received by connect callbacks.
 * @see - {@link WinccoaDpConnectCallback}
 * - {@link WinccoaDpQueryConnectCallback}
 */
var WinccoaConnectUpdateType;
(function (WinccoaConnectUpdateType) {
    // needs to be synchronized with NodeJsHotLinkWaitForAnswerBase.hxx
    /** Normal update after value changes. */
    WinccoaConnectUpdateType[WinccoaConnectUpdateType["Normal"] = 0] = "Normal";
    /** Initial update with current values. */
    WinccoaConnectUpdateType[WinccoaConnectUpdateType["Answer"] = 1] = "Answer";
    /** Refresh update after REDU switch or DIST connection. */
    WinccoaConnectUpdateType[WinccoaConnectUpdateType["Refresh"] = 2] = "Refresh";
})(WinccoaConnectUpdateType || (exports.WinccoaConnectUpdateType = WinccoaConnectUpdateType = {}));
/**
 * Enumerates the security events that can be reported from this manager.
 * @see - {@link WinccoaManager.securityEvent}
 */
var WinccoaSecurityEventId;
(function (WinccoaSecurityEventId) {
    /** Unknown ID - do not use */
    WinccoaSecurityEventId[WinccoaSecurityEventId["Unknown"] = 0] = "Unknown";
    /** Server port has been opened. Required additional parameters when calling
     *  {@link WinccoaManager.securityEvent}:
     * - args[0] - port: number
     * - args[1] - protocolDetails: string (e.g. `'https://'`)
     */
    WinccoaSecurityEventId[WinccoaSecurityEventId["PortOpened"] = 1] = "PortOpened";
})(WinccoaSecurityEventId || (exports.WinccoaSecurityEventId = WinccoaSecurityEventId = {}));
/**
 * Enumerates the type of a data point element, as returned by
 * {@link WinccoaManager.dpElementType}.
 * @see {@link WinccoaManager.dpElementType}.
 */
var WinccoaElementType;
(function (WinccoaElementType) {
    /** bit pattern */
    WinccoaElementType[WinccoaElementType["Bit32"] = 24] = "Bit32";
    /** bit pattern structure */
    WinccoaElementType[WinccoaElementType["Bit32Struct"] = 16] = "Bit32Struct";
    /** bit pattern */
    WinccoaElementType[WinccoaElementType["Bit64"] = 50] = "Bit64";
    /** bit pattern structure */
    WinccoaElementType[WinccoaElementType["Bit64Struct"] = 52] = "Bit64Struct";
    /** blob (binary large object) */
    WinccoaElementType[WinccoaElementType["Blob"] = 46] = "Blob";
    /** blob structure */
    WinccoaElementType[WinccoaElementType["BlobStruct"] = 47] = "BlobStruct";
    /** bit */
    WinccoaElementType[WinccoaElementType["Bool"] = 23] = "Bool";
    /** bit structure */
    WinccoaElementType[WinccoaElementType["BoolStruct"] = 15] = "BoolStruct";
    /** character */
    WinccoaElementType[WinccoaElementType["Char"] = 19] = "Char";
    /** character structure */
    WinccoaElementType[WinccoaElementType["CharStruct"] = 11] = "CharStruct";
    /** DP-Identifier */
    WinccoaElementType[WinccoaElementType["Dpid"] = 27] = "Dpid";
    /** label structure */
    WinccoaElementType[WinccoaElementType["DpidStruct"] = 39] = "DpidStruct";
    /** dynamic array for bit pattern */
    WinccoaElementType[WinccoaElementType["DynBit32"] = 8] = "DynBit32";
    /** dynamic bit-pattern array structure */
    WinccoaElementType[WinccoaElementType["DynBit32Struct"] = 35] = "DynBit32Struct";
    /** dynamic array for bit pattern */
    WinccoaElementType[WinccoaElementType["DynBit64"] = 51] = "DynBit64";
    /** dynamic bit-pattern array structure */
    WinccoaElementType[WinccoaElementType["DynBit64Struct"] = 53] = "DynBit64Struct";
    /** dynamic blob */
    WinccoaElementType[WinccoaElementType["DynBlob"] = 48] = "DynBlob";
    /** dynamic blob structure */
    WinccoaElementType[WinccoaElementType["DynBlobStruct"] = 49] = "DynBlobStruct";
    /** dynamic bit array */
    WinccoaElementType[WinccoaElementType["DynBool"] = 7] = "DynBool";
    /** dynamic bit array structure */
    WinccoaElementType[WinccoaElementType["DynBoolStruct"] = 34] = "DynBoolStruct";
    /** dynamic character array */
    WinccoaElementType[WinccoaElementType["DynChar"] = 3] = "DynChar";
    /** dynamic character array structure */
    WinccoaElementType[WinccoaElementType["DynCharStruct"] = 30] = "DynCharStruct";
    /** dynamic DP-Identifier array */
    WinccoaElementType[WinccoaElementType["DynDpid"] = 29] = "DynDpid";
    /** dynamic DP-Identifier array structure */
    WinccoaElementType[WinccoaElementType["DynDpidStruct"] = 38] = "DynDpidStruct";
    /** dynamic number array for floating decimal point */
    WinccoaElementType[WinccoaElementType["DynFloat"] = 6] = "DynFloat";
    /** dynamic number structure for floating decimal point */
    WinccoaElementType[WinccoaElementType["DynFloatStruct"] = 33] = "DynFloatStruct";
    /** dynamic integer array */
    WinccoaElementType[WinccoaElementType["DynInt"] = 5] = "DynInt";
    /** dynamic integer structure */
    WinccoaElementType[WinccoaElementType["DynIntStruct"] = 32] = "DynIntStruct";
    /** multilingual dynamic text array */
    WinccoaElementType[WinccoaElementType["DynLangString"] = 44] = "DynLangString";
    /** multilingual dynamic text structure */
    WinccoaElementType[WinccoaElementType["DynLangStringStruct"] = 45] = "DynLangStringStruct";
    /** dynamic array of Integer values (64 bit) */
    WinccoaElementType[WinccoaElementType["DynLong"] = 55] = "DynLong";
    /** dynamic number structure for integer value (64 bit) */
    WinccoaElementType[WinccoaElementType["DynLongStruct"] = 57] = "DynLongStruct";
    /** dynamic text array */
    WinccoaElementType[WinccoaElementType["DynString"] = 9] = "DynString";
    /** dynamic text-array structure */
    WinccoaElementType[WinccoaElementType["DynStringStruct"] = 36] = "DynStringStruct";
    /** dynamic time array */
    WinccoaElementType[WinccoaElementType["DynTime"] = 10] = "DynTime";
    /** dynamic time array structure */
    WinccoaElementType[WinccoaElementType["DynTimeStruct"] = 37] = "DynTimeStruct";
    /** dynamic array of positive whole numbers */
    WinccoaElementType[WinccoaElementType["DynUInt"] = 4] = "DynUInt";
    /** dynamic array of positive integers */
    WinccoaElementType[WinccoaElementType["DynUIntStruct"] = 31] = "DynUIntStruct";
    /** dynamic array of Positive integer values (64 bit) */
    WinccoaElementType[WinccoaElementType["DynULong"] = 59] = "DynULong";
    /** dynamic number structure for positive integer value (64 bit) */
    WinccoaElementType[WinccoaElementType["DynULongStruct"] = 61] = "DynULongStruct";
    /** floating point system */
    WinccoaElementType[WinccoaElementType["Float"] = 22] = "Float";
    /** number structure for floating decimal point */
    WinccoaElementType[WinccoaElementType["FloatStruct"] = 14] = "FloatStruct";
    /** integer */
    WinccoaElementType[WinccoaElementType["Int"] = 21] = "Int";
    /** integer structure */
    WinccoaElementType[WinccoaElementType["IntStruct"] = 13] = "IntStruct";
    /** description */
    WinccoaElementType[WinccoaElementType["LangString"] = 42] = "LangString";
    /** description structure */
    WinccoaElementType[WinccoaElementType["LangStringStruct"] = 43] = "LangStringStruct";
    /** Integer value (64 bit) */
    WinccoaElementType[WinccoaElementType["Long"] = 54] = "Long";
    /** structure for integer value (64 bit) */
    WinccoaElementType[WinccoaElementType["LongStruct"] = 56] = "LongStruct";
    /** text */
    WinccoaElementType[WinccoaElementType["String"] = 25] = "String";
    /** text structure */
    WinccoaElementType[WinccoaElementType["StringStruct"] = 17] = "StringStruct";
    /** structure */
    WinccoaElementType[WinccoaElementType["Struct"] = 1] = "Struct";
    /** time */
    WinccoaElementType[WinccoaElementType["Time"] = 26] = "Time";
    /** time structure */
    WinccoaElementType[WinccoaElementType["TimeStruct"] = 18] = "TimeStruct";
    /** data point type reference */
    WinccoaElementType[WinccoaElementType["Typeref"] = 41] = "Typeref";
    /** unsigned integer */
    WinccoaElementType[WinccoaElementType["UInt"] = 20] = "UInt";
    /** structure of unsigned integers */
    WinccoaElementType[WinccoaElementType["UIntStruct"] = 12] = "UIntStruct";
    /** Positive integer value (64 bit) */
    WinccoaElementType[WinccoaElementType["ULong"] = 58] = "ULong";
    /** structure for positive integer value (64 bit) */
    WinccoaElementType[WinccoaElementType["ULongStruct"] = 60] = "ULongStruct";
})(WinccoaElementType || (exports.WinccoaElementType = WinccoaElementType = {}));
/**
 * Main class of the TypeScript/JavaScript API to the __WinCC OA JavaScript Manager for Node.js&#174;__.
 *
 * > **IMPORTANT**
 * >
 * > All methods described in this documentation must be __called from code inside a
 * > method or function__ to prevent unexpected or undefined behavior.
 *
 * To use the API, an instance of this class needs to be created, e. g.:
 * @example
 * ```ts
 * import { WinccoaManager } from 'winccoa-manager';
 * const winccoa = new WinccoaManager();
 * ```
 */
class WinccoaManager {
    /**
     * Constructor - creates new instance of an API object.
     * @example
     * ```ts
     * import { WinccoaManager } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     * ```
     */
    constructor() {
        this.api = new winccoaconnection_1.WinccoaManagerApi();
        this.winccoaManagerConnection = connection_binding_1.ConnectionBinding.getInstance();
    }
    /**
     * Writes an information log entry. `console.log()` and `console.info()` are mapped
     * to this method.
     * @param args Information to be written to log.
     * @example
     * ```ts
     * let note = "Additional note.";
     * winccoa.logInfo("Info message", note);
     * ```
     */
    logInfo(...args) {
        this.winccoaManagerConnection.log.info(...args);
    }
    /**
     * Writes a warning log entry. `console.warn()` is mapped to this method.
     * @param args Warning information to be written to log.
     * @example
     * ```ts
     * let note = "Additional note.";
     * winccoa.logWarning("Warning message", note);
     * ```
     */
    logWarning(...args) {
        this.winccoaManagerConnection.log.warning(...args);
    }
    /**
     * Writes a severe error log entry. `console.error()` is mapped to this method.
     * @param args Severe error information to be written to log.
     * @example
     * ```ts
     * let note = "Additional note.";
     * winccoa.logSevere("Severe message", note);
     * ```
     */
    logSevere(...args) {
        this.winccoaManagerConnection.log.severe(...args);
    }
    /**
     * Writes a fatal error log entry. This also exits the WinCC OA manager.
     * @param args Fatal error information to be written to log.
     * @example
     * ```ts
     * let note = "Additional note.";
     * winccoa.logFatal("Fatal message", note);
     * ```
     */
    logFatal(...args) {
        this.winccoaManagerConnection.log.fatal(...args);
    }
    /**
     * Returns list of project, sub-project and product installation paths.
     * @returns List of project, sub-project and product installation paths.
     * The project path is always a first entry and the installation path is always the last.
     * @example
     * ```ts
     * let paths = winccoa.getPaths();
     * for (let i = 0; i < paths.length; i++) {
     *   console.info(paths.at(i));
     * }
     * ```
     */
    getPaths() {
        return this.winccoaManagerConnection.paths;
    }
    /**
     * Exits the WinCC OA manager.
     * @param exitCode Exit code to return to the operating system.
     * @example
     * ```ts
     * winccoa.exit();
     * ```
     */
    exit(exitCode = 0) {
        this.winccoaManagerConnection.exit(exitCode);
    }
    /**
     * Returns all available options and their values.
     * The options can be set with {@link setOptions}.
     * @returns Object containing all current option values.
     * @see - {@link setOptions}
     * - {@link WinccoaOptions}
     * @example
     * ```ts
     * let options = winccoa.getOptions();
     * for (const [key, value] of Object.entries(options)) {
     *   console.info(`Name: ${key} - Value: ${value}`);
     * }
     * ```
     */
    getOptions() {
        return this.api.getOptions();
    }
    /**
     * Set one or more options.
     * The options and their values can be retrieved with {@link getOptions}.
     * @note The userId option cannot be set with this method.
     * <br>The user can be set with {@link setUserId} for the instance of `WinccoaManager`.
     * <br>The user can be set with `-user` manager option for the node manager,
     * see [Administration of managers](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Pmon_Consolepanel/Pmon_Consolepanel-23.html).
     * @param obj The options to be set (see {@link WinccoaOptions} for possible options).
     * @returns Boolean `true` on success, otherwise {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} if option has wrong type or option value is out of range.
     * @see - {@link getOptions}
     * - {@link WinccoaOptions}
     * - {@link setUserId}
     * - [Administration of managers](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Pmon_Consolepanel/Pmon_Consolepanel-23.html)
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaLangTextFormat } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * function setOptionsTest() {
     *   try {
     *     winccoa.setOptions({
     *       langTextFormat: WinccoaLangTextFormat.StringFixed,
     *       langIdx: 1,
     *     });
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     * ```
     */
    setOptions(obj) {
        return this.api.setOptions(obj);
    }
    /**
     * Returns information about current API and WinCC OA versions.
     * */
    getVersionInfo() {
        return { api: { version: 1 }, winccoa: this.api.getOaVersionDetails() };
    }
    /**
     * Checks the existence of a valid data point identifier.
     * @param dpeName A data point identifier: a sys, a DPT, a DP, a DPE, a config, a detail or an attr.
     * @returns true if at least one part of a data point identifier can be resolved correctly, otherwise false.
     * @throws {@link WinccoaError} when invalid argument type is given.
     * @see - [CTRL function `dpExists()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpExists.html)
     * @example
     * ```ts
     * let isDpExists = false;
     * try {
     *   isDpExists = winccoa.dpExists('ExampleDP_Arg1.');
     * } catch (exc) {
     *   console.error(exc);
     * }
     * console.info("Is ExampleDP_Arg1 exists - " + isDpExists);
     * ```
     */
    dpExists(dpeName) {
        return this.api.dpExists(dpeName);
    }
    /**
     * Returns all the data point names or the data point element names that match a pattern in alphabetical order.
     * @param dpPattern Serach pattern. When an empty pattern is given (=default), then returns all datapoints.
     * <br>Wildcards are used to filter data point name.
     * The charcters `*` and `?` are used for the purpose,
     * where the asterisk (`*`) replaces any number of characters and the question mark `?` stands for just one character.
     * Only data points that have the same number of levels as specified are found.
     * Levels are separated by a period. `dpNames(**)` is equivalent to `dpNames(*.*)`.
     * <br>Furthermore:
     * - `:*` returns all configs, `:config.*` returns all details, `:config.detail.*` returns all attributes
     * - `dp.el:*` returns only the configs according to the DPE. , for example, no `_original` for a node.
     *
     * <br>Wildcards can be used in arrays (square brackets , e.g.: `[0,3,5-7]` - numbers 0,3,5,6,7) or outside arrays in option lists (in curly brackets `{}`).
     * <br>Example of wildcards in lists of options:
     * ```
     * winccoa.dpNames('{*.Ala.*,*.Ala*}', dpType);
     * winccoa.dpNames('*{.Ala.,.Ala}*', dpType);
     * winccoa.dpNames('*.A{la.,la}*', dpType);
     * winccoa.dpNames('*.Al{a.,a}*', dpType);
     * winccoa.dpNames('*.Ala{.,}*', dpType);
     * winccoa.dpNames('*.Ala{.}*', dpType);
     * winccoa.dpNames('*.Ala.*', dpType);
     * winccoa.dpNames('*.Ala*', dpType);
     * ```
     * @param dpType Data point type. Allows to restrict the returned data points to a specific data point type.
     * When using the parameter only data points that match the pattern and the selected data point type will be returned.
     * @param ignoreCase Defines if the search should ignore the casing of the search pattern (=true) or not (=false, default)
     * @returns List with data points or data point element names.
     * @throws {@link WinccoaError} when invalid argument type is given.
     * @see - [CTRL function `dpNames()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpNames.html)
     * @example
     * ```ts
     * let foundDpNames: string[] = [];
     * try {
     *   foundDpNames = winccoa.dpNames('ExampleDP*', 'ExampleDP_Float');
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * for (let i = 0; i < foundDpNames.length; i++) {
     *   console.info("DPE name: " + foundDpNames.at(i));
     * }
     * ```
     */
    dpNames(dpPattern = '', dpType = '', ignoreCase = false) {
        return this.api.dpNames(dpPattern, dpType, ignoreCase).sort();
    }
    /**
     * Returns all or selected data point types from the current project.
     * @param pattern Pattern for the returned DPTs. When an empty pattern is given (=default), then returns all DP types.
     * <br>Wildcards are used to filter data point type name.
     * The charcters `*` and `?` are used for the purpose,
     * where the asterisk (`*`) replaces any number of characters and the question mark `?` stands for just one character.
     * <br>Wildcards can be used in arrays (square brackets , e.g.: `[0,3,5-7]` - numbers 0,3,5,6,7) or outside arrays in option lists (in curly brackets `{}`).
     * <br>Example of wildcards in lists of options:
     * ```
     * winccoa.dpTypes('{*.Ala.*,*.Ala*}');
     * winccoa.dpTypes('*{.Ala.,.Ala}*');
     * winccoa.dpTypes('*.A{la.,la}*');
     * winccoa.dpTypes('*.Al{a.,a}*');
     * winccoa.dpTypes('*.Ala{.,}*');
     * winccoa.dpTypes('*.Ala{.}*');
     * winccoa.dpTypes('*.Ala.*');
     * winccoa.dpTypes('*.Ala*');
     * ```
     * @param systemId The desired system if querying from other systems. Optional parameter.
     *                 If this parameter is not defined, the own system is queried.
     * @param includeEmpty When this is set to false, data point types without existing
     *                     data points will be ignored.
     * @returns String array with all DP type names.
     * @throws {@link WinccoaError} when invalid argument type or non-existing __systemId__ is given.
     * @see - [CTRL function `dpTypes()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpTypes.html)
     * @example
     * ```ts
     * try {
     *   const foundDpTypes = winccoa.dpTypes('ExampleDP*', 1, true);
     *   for (let i = 0; i < foundDpTypes.length; i++) {
     *     console.info("DP type: " + foundDpTypes.at(i));
     *   }
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    dpTypes(pattern = '', systemId = -1, includeEmpty = true) {
        return this.api.dpTypes(pattern, systemId, includeEmpty);
    }
    /**
     * Search for file or directory in WinCC OA project and installation paths.
     * @param fileDirName the file or directory to search for
     * @returns full path of the file or directory or an empty string if not found
     * @throws {@link WinccoaError} when invalid argument type is given.
     * @example
     * ```ts
     * let configPath;
     * try {
     *   configPath = winccoa.findFile('config/config');
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info("Config file full path: " + configPath);
     * ```
     */
    findFile(fileDirName) {
        return this.winccoaManagerConnection.findFile(fileDirName);
    }
    //------------------------------------------------------------------------
    /**
     * Creates a data point.
     * @param dpeName Name of the data point to be created.
     * @param dpType Type of data point to be created.
     * @param systemId To create a data point on a remote system in a distributed system,
     *                 this parameter must contain the system number.
     * @param dpId The ID of the data point. If a data point with the given ID already
     *             exists, a random ID is chosen.
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     * @throws {@link WinccoaError} in case of:
     * - invalid argument type is given,
     * - invalid __dpeName__, __dpType__ or non-existing __systemId__ is given,
     * - data point with the given __dpeName__ is already exist.
     * @see - {@link dpDelete}
     * - [CTRL function `dpCreate()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpCreate.html)
     * @example
     * ```ts
     * let dpCreated = false;
     * try {
     *   dpCreated = await winccoa.dpCreate('newFloatDpe', 'ExampleDP_Float');
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info("DP newFloatDpe is created - " + dpCreated);
     * ```
     */
    dpCreate(dpeName, dpType, systemId, dpId) {
        return this.api.dpCreate(dpeName, dpType, systemId, dpId);
    }
    //------------------------------------------------------------------------
    /**
     * Set the value of one or more datapoint element(s).
     * @param dpeNames Datapoint element name(s) of the values to set.
     * @param values Values to set. Must have the same size as __dpeNames__. If __dpeNames__ is a
     *               single string and not an array, this parameter must also be a single value
     *               and not an array.
     * @returns `true` if successful, otherwise {@link WinccoaError} wil be thrown instead.
     * > Since this method does not wait for the actual value update
     * > in the database, __the update can still fail__ after this method returned `true`. Use
     * {@link dpSetWait} to also get informed about these errors.
     * @throws {@link WinccoaError} when DPE names do not exist, values cannot be converted, array
     *         sizes mismatch etc.
     * @see - {@link dpSetWait}
     * - {@link dpSetTimed}
     * - {@link dpSetTimedWait}
     * - [CTRL function `dpSet()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSet.html)
     * @example
     * ```ts
     * const dpes = ['ExampleDP_Arg1.', 'ExampleDP_DDE.b1'];
     * const values = [123.456, false];
     * try {
     *   winccoa.dpSet(dpes, values); // Two arrays of size 2
     *   winccoa.dpSet('ExampleDP_Arg2.', 2); // single value
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    dpSet(dpeNames, values) {
        if (Array.isArray(dpeNames)) {
            return this.api.dpSet(dpeNames, values);
        }
        else {
            return this.api.dpSet([dpeNames], [values]);
        }
    }
    //------------------------------------------------------------------------
    /**
     * {@inheritDoc dpSet}
     * @returns Promise that resolves to `true` if succesful. If not successful,
     *          a {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when DPE names do not exist, values cannot be converted, array
     *         sizes mismatch, no write access etc.
     * @see - {@link dpSet}
     * - {@link dpSetTimed}
     * - {@link dpSetTimedWait}
     * - [CTRL function `dpSetWait()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetWait.html)
     * @example
     * ```ts
     * const dpes = ['ExampleDP_Arg1.', 'ExampleDP_DDE.b1'];
     * const values = [123.456, false];
     * try {
     *   await winccoa.dpSetWait(dpes, values); // Two arrays of size 2
     *   await winccoa.dpSetWait('ExampleDP_Arg2.', 2); // single value
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    dpSetWait(dpeNames, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(dpeNames))
            return this.api.dpSetWait(dpeNames, values);
        return this.api.dpSetWait([dpeNames], [values]);
    }
    //------------------------------------------------------------------------
    /**
     * Set values of one or more datapoint element(s) with a given source time.
     * @param time Source time for the value change.
     * @param dpeNames Datapoint element name(s) of the values to set.
     * @param values Values to set. Must have the same size as __dpeNames__. If __dpeNames__ is a
     *               single string and not an array, this parameter must also be a single value
     *               and not an array.
     * @returns Boolean `true` in case of a success, otherwise `false`.
     * > Since this method does not wait for the actual value update
     * > in the database, __the update can still fail__ after this method returned `true`. Use
     * {@link dpSetTimedWait} to also get informed about these errors.
     * @throws {@link WinccoaError} when DPE names do not exist, values cannot be converted, array
     *         sizes mismatch etc.
     * @see - {@link dpSet}
     * - {@link dpSetWait}
     * - {@link dpSetTimedWait}
     * - [CTRL function `dpSetTimed()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetTimed.html)
     * @example
     * ```ts
     * const timeStamp = new Date('2023-01-03T04:05:06.789Z');
     * let isSuccess = false;
     * try {
     *   isSuccess = winccoa.dpSetTimed(timeStamp, 'ExampleDP_Arg1.', 2);
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info("dpSetTimed call is successed - " + isSuccess);
     * ```
     */
    dpSetTimed(time, dpeNames, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(dpeNames)) {
            return this.api.dpSetTimed(time, dpeNames, values);
        }
        else {
            return this.api.dpSetTimed(time, [dpeNames], [values]);
        }
    }
    //------------------------------------------------------------------------
    /**
     * {@inheritDoc dpSetTimed}
     * @returns Promise that resolves to `true` if succesful. If not successful,
     *          a {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when DPE names do not exist, values cannot be converted, array
     *         sizes mismatch, no write access etc.
     * @see - {@link dpSet}
     * - {@link dpSetTimed}
     * - {@link dpSetWait}
     * - [CTRL function `dpSetTimedWait()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetTimedWait.html)
     * @example
     * ```ts
     * const timeStamp = new Date('2023-01-03T04:05:06.789Z');
     * let isSuccess = false;
     * try {
     *   isSuccess = await winccoa.dpSetTimedWait(timeStamp, 'ExampleDP_Arg1.', 2);
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info("Time and value are set for ExampleDP_Arg1 - " + isSuccess);
     * ```
     */
    dpSetTimedWait(time, dpeNames, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(dpeNames))
            return this.api.dpSetTimedWait(time, dpeNames, values);
        return this.api.dpSetTimedWait(time, [dpeNames], [values]);
    }
    //------------------------------------------------------------------------
    /**
     * Get the current values of one or more datapoint elements.
     * @async
     * @param dpeNames Datapoint element name(s) of the values to get.
     * @returns Promise that resolves to the current value(s) of the DPE(s). The received values
     *          must be cast to their expected types before they can be used.
     * @throws {@link WinccoaError} when DPE does not exist or current user has no read access to it.
     * @see [CTRL function `dpGet()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpGet.html)
     * @example
     * ```ts
     * const dpes = ['ExampleDP_Arg1.', 'ExampleDP_DDE.b1'];
     * try {
     *   const values = (await winccoa.dpGet(dpes)) as [number, boolean];
     *   for (let i = 0; i < dpes.length; i++) {
     *     console.info(`${dpes[i]}: ${values[i]}`);
     *   }
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    dpGet(dpeNames) {
        if (Array.isArray(dpeNames))
            return this.api.dpGet(dpeNames);
        else
            return this.api
                .dpGet([dpeNames])
                .then((valueArray) => valueArray[0]);
    }
    //------------------------------------------------------------------------
    /**
     * Allows to set data point alert attributes.
     * @note The attributes and their constants which can be set with this method are described in the chapter
     * [_alert_hdl](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Notes/dpconfig_alert_hdl.html)
     * @param alerts Alert(s) of type {@link WinccoaAlertTime} to be set.
     * @param values Attribute value(s) to be set. Must have the same size as __alerts__. If __alerts__ is a
     *               single string and not an array, this parameter must also be a single value and not an array.
     * @returns Boolean `true` in case of a success, otherwise {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when DPE does not exist, mismatch number of DPs and values,
     * invalid argument type, etc.
     * @see - {@link alertSetWait}
     * - {@link alertSetTimed}
     * - {@link alertSetTimedWait}
     * - {@link WinccoaAlertTime}
     * - [_alert_hdl](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Notes/dpconfig_alert_hdl.html)
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaAlertTime } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * async function alertSetTest() {
     *   const dpeWithAlert = 'ExampleDP_AlertHdl1.';  // assuming alert is triggered
     *   const result = await winccoa.dpQuery(
     *     "SELECT ALERT '_alert_hdl.._act_state', '_alert_hdl.._value' FROM '" +
     *     dpeWithAlert +
     *     "'",
     *   );
     *
     *   const ts = result[result.length - 1][1] as WinccoaAlertTime;
     *   const alertTime = new WinccoaAlertTime(ts.date, ts.count, ts.dpe + '._comment');
     *   let isSuccess = false;
     *   try {
     *     isSuccess = winccoa.alertSet(alertTime, 'Alert comment text from winccoa node manager.');
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     *
     *   console.info("AlertSet is called successfully - " + isSuccess);
     * }
     * ```
     */
    alertSet(alerts, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(alerts))
            return this.api.alertSet(alerts, values);
        return this.api.alertSet([alerts], [values]);
    }
    //------------------------------------------------------------------------------
    /**
     * {@inheritDoc alertSet}
     * @returns Promise that resolves to `true` if successful. If not successful,
     *          a {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when DPE does not exist, mismatch number of DPs and values,
     * invalid argument type, etc.
     * @see - {@link alertSet}
     * - {@link alertSetTimed}
     * - {@link alertSetTimedWait}
     * - {@link WinccoaAlertTime}
     * - [_alert_hdl](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Notes/dpconfig_alert_hdl.html)
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaAlertTime } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * async function alertSetWaitTest() {
     *   const dpeWithAlert = 'ExampleDP_AlertHdl1.'; // assuming alert is triggered
     *   const result = await winccoa.dpQuery(
     *     `SELECT ALERT '_alert_hdl.._act_state', '_alert_hdl.._value' FROM '${dpeWithAlert}'`,
     *   );
     *
     *   const ts = result[result.length - 1][1] as WinccoaAlertTime;
     *   const alertTime = new WinccoaAlertTime(
     *     ts.date,
     *     ts.count,
     *     ts.dpe + '._comment',
     *   );
     *
     *   try {
     *     await winccoa.alertSetWait(
     *       alertTime,
     *       'Alert comment text from winccoa node manager.',
     *     );
     *     console.info(`Comment for alert of '${dpeWithAlert}' is successfully set`);
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     * ```
     */
    alertSetWait(alerts, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(alerts))
            return this.api.alertSetWait(alerts, values);
        return this.api.alertSetWait([alerts], [values]);
    }
    //------------------------------------------------------------------------------
    /**
     * Allows to set data point alert attributes with a given timestamp.
     * @note The attributes and their constants which can be set with this method are described in the chapter
     * [_alert_hdl](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Notes/dpconfig_alert_hdl.html)
     * @param time Source time for the attribute change.
     * @param alerts Alert(s) of type {@link WinccoaAlertTime} to be set.
     * @param values Attribute value to be set. Must have the same size as __alerts__. If __alerts__ is a
     *               single string and not an array, this parameter must also be a single value and not an array.
     * @returns Boolean `true` in case of a success, otherwise {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when DPE does not exist, mismatch number of DPs and values,
     * invalid argument type, etc.
     * @see - {@link alertSet}
     * - {@link alertSetWait}
     * - {@link alertSetTimedWait}
     * - {@link WinccoaAlertTime}
     * - [_alert_hdl](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Notes/dpconfig_alert_hdl.html)
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaAlertTime } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * async function alertSetTimedTest() {
     *   const dpeWithAlert = 'ExampleDP_AlertHdl1.';  // assuming alert is triggered
     *   const result = await winccoa.dpQuery(
     *     "SELECT ALERT '_alert_hdl.._act_state', '_alert_hdl.._value' FROM '" +
     *     dpeWithAlert +
     *     "'",
     *   );
     *
     *   const timeStamp = new Date('2024-03-19T04:05:06.789Z');
     *   const ts = result[result.length - 1][1] as WinccoaAlertTime;
     *   const alertTime = new WinccoaAlertTime(ts.date, ts.count, ts.dpe + '._ack_state');
     *   let isSuccess = false;
     *   try {
     *     isSuccess = winccoa.alertSetTimed(timeStamp, alertTime, 1);
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     *
     *   console.info("AlertSet is called successfully - " + isSuccess);
     * }
     * ```
     */
    alertSetTimed(time, alerts, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(alerts))
            return this.api.alertSetTimed(time, alerts, values);
        return this.api.alertSetTimed(time, [alerts], [values]);
    }
    //------------------------------------------------------------------------------
    /**
     * {@inheritDoc alertSetTimed}
     * @returns Promise that resolves to `true` if successful. If not successful,
     *          a {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when DPE does not exist, mismatch number of DPs and values,
     * invalid argument type, etc.
     * @see - {@link alertSet}
     * - {@link alertSetWait}
     * - {@link alertSetTimed}
     * - {@link WinccoaAlertTime}
     * - [_alert_hdl](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/Notes/dpconfig_alert_hdl.html)
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaAlertTime } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * async function alertSetTimedWaitTest() {
     *   const dpeWithAlert = 'ExampleDP_AlertHdl1.'; // assuming alert is triggered
     *   const result = await winccoa.dpQuery(
     *     `SELECT ALERT '_alert_hdl.._act_state', '_alert_hdl.._value' FROM '${dpeWithAlert}'`,
     *   );
     *
     *   const timeStamp = new Date('2024-03-19T04:05:06.789Z');
     *   const ts = result[result.length - 1][1] as WinccoaAlertTime;
     *   const alertTime = new WinccoaAlertTime(
     *     ts.date,
     *     ts.count,
     *     ts.dpe + '._ack_state',
     *   );
     *
     *   try {
     *     await winccoa.alertSetTimedWait(timeStamp, alertTime, 1);
     *     console.info(`Alert is acknowledged at ${timeStamp}`);
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     * ```
     */
    alertSetTimedWait(time, alerts, 
    // unknown | unknown[] stays in the code to illustrate both uses in generated documentation
    // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
    values) {
        if (Array.isArray(alerts))
            return this.api.alertSetTimedWait(time, alerts, values);
        return this.api.alertSetTimedWait(time, [alerts], [values]);
    }
    //------------------------------------------------------------------------------
    /**
     * Creates a connection for being notified of datapoint element updates.
     * @param callback Function that is called whenever a connected value changes.
     * @param dpeNames DPE name(s) to connect to.
     * > - Each update will contain updates for all elements in __dpeNames__, not only
     *   the changed values.
     *   - Each update will contain an array of values, also if only a single datapoint element
     *     name is given.
     * @param answer if `true`, __callback__ is called with the initial values right away, if `false`,
     *               callback is only called after an actual value change.
     * @throws {@link WinccoaError} when invalid parameter types, unknown datapoint element names etc.
     * @returns ID of the new connection (>= 0). This can be used to disconnect from
     *          updates if required with {@link dpDisconnect}. Otherwise the connection will be closed
     *          when the manager exits.
     * @remark For an example describing how to pass user data to a callback, see the second example
     *         for {@link WinccoaDpQueryConnectCallback}.
     * @see - {@link dpDisconnect}
     * - {@link WinccoaDpConnectCallback}
     * - {@link WinccoaConnectUpdateType}
     * @example
     * ``` ts
     * import {
     *   WinccoaManager,
     *   WinccoaConnectUpdateType,
     *   WinccoaError
     * } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * function connectCB(
     *   names: string[],
     *   values: unknown[],
     *   type: WinccoaConnectUpdateType,
     *   error?: WinccoaError
     * ) {
     *   if (error) {
     *     console.log(error);
     *     return;
     *   }
     *
     *   if (type == WinccoaConnectUpdateType.Answer)
     *     console.warn('--- Initial update ---');
     *
     *   for (let i = 0; i < names.length; i++)
     *     console.info(`[${i}] '${names[i]}' : ${values[i]}`);
     * }
     *
     * function connect() {
     *   let id = -1;
     *   try {
     *     id = winccoa.dpConnect(
     *       connectCB,
     *       ['ExampleDP_Arg1.', 'ExampleDP_Arg2.'],
     *       true
     *     );
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     * ```
     */
    dpConnect(callback, dpeNames, answer = true) {
        if (Array.isArray(dpeNames)) {
            return this.api.dpConnect(callback, dpeNames, answer);
        }
        else {
            return this.api.dpConnect(callback, [dpeNames], answer);
        }
    }
    //------------------------------------------------------------------------
    /**
     * Disconnect from datapoint element update connections established with {@link dpConnect}.
     *
     * @param id ID of the connection to close as returned by {@link dpConnect}.
     * @throws {@link WinccoaError} when __id__ is not found or invalid.
     * @returns ID of the connection that has been closed (>= 0).
     * @see - {@link dpConnect}
     * @example
     * ```ts
     * function connect() {
     *   let id = -1;
     *   try {
     *     id = winccoa.dpConnect(
     *       connectCB,
     *       ['ExampleDP_Arg1.', 'ExampleDP_Arg2.'],
     *       true
     *     );
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     *
     * function disconnect() {
     *   try {
     *     winccoa.dpDisconnect(id);
     *   } catch (exc) {
     *    console.error(exc);
     *   }
     * }
     * ```
     */
    dpDisconnect(id) {
        return this.api.dpDisconnect(id);
    }
    //------------------------------------------------------------------------
    /**
     * Returns the system ID.
     * @param systemName - The name of the system (optional).
     * If it is not given, then the system ID of its own system will be returned.
     * @returns The system ID.
     * @throws {@link WinccoaError} when invalid system name is given.
     * @see - {@link getSystemName}
     * - [CTRL function `getSystemId()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlE_R/getSystemId.html)
     * @example
     * ```ts
     * let ownId;
     * let system1Id;
     * try {
     *   ownId = winccoa.getSystemId();
     *   system1Id = winccoa.getSystemId('System1:');
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info("Own system id = " + ownId);
     * console.info("System1 id = " + system1Id);
     * ```
     */
    getSystemId(systemName) {
        return this.api.getSystemId(systemName);
    }
    //------------------------------------------------------------------------
    /**
     * Returns the system name.
     * @param systemId - System ID (optional).
     * If it is not given, then the system name of its own system will be returned.
     * @returns The system name.
     * @throws {@link WinccoaError} when invalid system ID is given.
     * @see - {@link getSystemId}
     * - [CTRL function `getSystemName()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlE_R/getSystemName.html)
     * @example
     * ```ts
     * try {
     *   const ownName = winccoa.getSystemName();
     *   const id1Name = winccoa.getSystemName(1);
     *   console.info(
     *     `Own system name = '${ownName}, system with id 1 name = '${id1Name}'`
     *   );
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    getSystemName(systemId) {
        return this.api.getSystemName(systemId);
    }
    //------------------------------------------------------------------------
    /**
     * Deletes a data point.
     * @param dpName - Name of the data point to be deleted.
     *            In case of a distributed system the name of the data point
     *            to be deleted must contain the system name.
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     * @throws {@link WinccoaError} when datapoint with the given name does not exist or current user has no privileges to delete a DP.
     * @see - {@link dpCreate}
     * - [CTRL function `dpDelete()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpDelete.html)
     * @example
     * ```ts
     * let isSuccess = false;
     * try {
     *   isSuccess = await winccoa.dpCreate('newDpe', 'ExampleDP_Float');
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * if (isSuccess)
     * {
     *   try {
     *     isSuccess = await winccoa.dpDelete('newDpe');
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     *
     * console.info("DP newDpe was deleted successfully - " + isSuccess);
     * ```
     */
    dpDelete(dpName) {
        return this.api.dpDelete(dpName);
    }
    //------------------------------------------------------------------------
    /**
     * Copies a data point including its configuration.
     * @param source Name of the datapoint to copy.
     * @param destination Name of the new copied datapoint. Must not exist yet.
     * @param driver Optional driver number (default 1).
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     *          In case of an error, `error.details` will contain the same error code
     *          that CTRL function dpCopy() would return (see there).
     * @throws {@link WinccoaError} when given source datapoint does not exist, when datapoint is copied into itself, invalid argument given, etc.
     * @see - [CTRL function `dpCopy()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpCopy.html)
     * @example
     * ```ts
     * async function dpCopyTest() {
     *   let isSuccess = false;
     *   try {
     *     isSuccess = await winccoa.dpCopy('ExampleDP_Arg1', 'ExampleDP_Arg3');
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     *
     *   console.info("DP ExampleDP_Arg1 is copied to ExampleDP_Arg3 successfully - " + isSuccess);
     * }
     * ```
     */
    dpCopy(source, destination, driver) {
        return this.api.dpCopy(source, destination, driver).then(
        // dpCopy just returns an integer error code, make sure that errors are reported
        // as exceptions and true is returned when successful.
        (errorCode) => {
            if (errorCode != 0)
                // Create an error that mimics the errors created in C++
                throw new error_1.WinccoaError(11, 'dpCopy() failed with error code', 'javascript', errorCode);
            return errorCode == 0;
        });
    }
    //------------------------------------------------------------------------
    /**
     * Retrieves attribute values with the help of SQL statements.
     * @param query - SQL statement.
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     * @note The query result has a table-like structure:
     * [0][0] (empty)    | [0][1] column header   |         ...
     * ----------------- | ---------------------- | ----------------------
     * [1][0] line name  | [1][0] content of line |         ...
     * [2][0] line name  | [2][1] content of line |         ...
     * ...               | ...                    |         ...
     *
     * e.g. this is the output for the query `"SELECT '_original.._value' FROM 'ExampleDP_Arg*'"` converted
     * to JSON:
     * ```
     * [
     *   ["",":_original.._value"],
     *   ["System1:ExampleDP_Arg1.",2.43],
     *   ["System1:ExampleDP_Arg2.",5.76]
     * ]
     * ```
     * @throws {@link WinccoaError} when invalid parameter or query string is given.
     * @see - {@link dpQueryConnectSingle}
     * - {@link dpQueryConnectAll}
     * - [CTRL function `dpQuery()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpQuery.html)
     * @example
     * ```ts
     * let result;
     * try {
     *   result = await winccoa.dpQuery(
     *     `SELECT '_original.._stime', '_original.._value' FROM 'ExampleDP_Arg*'`,
     *   );
     * } catch (exc) {
     *   console.error(exc);
     *   result = null;
     * }
     * if (result) {
     *   const queryTable: string[][] = result as string[][];
     *   for (let i = 0; i < queryTable.length; i++) {
     *     console.info(
     *       `query line - name: '%s' - timestamp = %s - value = %s`,
     *       ...queryTable[i],
     *     );
     *   }
     * }
     *  ```
     */
    dpQuery(query) {
        return this.api.dpQuery(query);
    }
    //------------------------------------------------------------------------
    /**
     * Calls __callback__ whenever one or more DPEs which meet the query condition are changed.
     * @param callback Function that is called whenever a subscribed DPE value changes.
     *                 The update message will contain only changed DPEs.
     * @param answer if true, callback is called with the initial DPE values right away, if false,
     *               callback is only called for an actual value change.
     * @param query - query as an SQL statement.
     * @throws {@link WinccoaError} when invalid parameter types, empty query, etc.
     * > When the query passed to this method is invalid, no exception is thrown,
     * > but the first (and only) callback will contain a {@link WinccoaError} instead.
     * @returns ID of the new connection (>= 0). This can be used to disconnect from
     *          updates if required with {@link dpQueryDisconnect}. Otherwise the connection will be closed
     *          when the manager exits.
     * @remark For an example describing how to pass user data to a callback, see the second example
     *         for {@link WinccoaDpQueryConnectCallback}.
     * @see - {@link dpQuery}
     * - {@link dpQueryConnectAll}
     * - {@link dpQueryDisconnect}
     * - {@link WinccoaConnectUpdateType}
     * - [CTRL function `dpQueryConnectSingle()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpQueryConnectSingle.html)
     * @example
     * ```ts
     * import {
     *   WinccoaManager,
     *   WinccoaConnectUpdateType,
     *   WinccoaError
     * } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * function  queryConnectCB (
     *   values: unknown[][],
     *   type: WinccoaConnectUpdateType,
     *   error?: WinccoaError
     * ) {
     *   if (error) {
     *     console.log(error);
     *     return;
     *   }
     *
     *   if (type == WinccoaConnectUpdateType.Answer)
     *     console.warn('--- Initial update ---');
     *
     *   for (let i = 0; i < values.length; i++) {
     *     console.info(`DPE = '%s', value = %s`, ...values[i]);
     *   }
     * }
     *
     * function connect() {
     *   let id = -1;
     *   try {
     *     id = winccoa.dpQueryConnectSingle(
     *       queryConnectCB,
     *       true,
     *       `SELECT '_online.._value' FROM '*' WHERE _DPT="ExampleDP_Float"`
     *     );
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     *  ```
     */
    dpQueryConnectSingle(callback, answer, query) {
        return this.api.dpQueryConnectSingle(callback, answer, query);
    }
    //------------------------------------------------------------------------
    /**
     * Calls __callback__ whenever one or more DPEs which meet the query condition change.
     * > It is recommended to use {@link dpQueryConnectSingle} if possible from the
     * > performance point of view.
     * @param callback Function that is called whenever a subscribed DPE value changes.
     *                 The update message will contain all subscribed DPEs.
     * @param answer if true, callback is called with the initial DPE values right away, if false,
     *               callback is only called for an actual value change.
     * @param query - query as an SQL statement.
     * @throws {@link WinccoaError} when invalid parameter types, empty or invalid query, etc.
     * > When the query passed to this method is invalid, no exception is thrown,
     * > but the first (and only) callback will contain a {@link WinccoaError} instead.
     * @returns ID of the new connection (>= 0). This can be used to disconnect from
     *          updates if required with {@link dpQueryDisconnect}. Otherwise the connection will be closed
     *          when the manager exits.
     * @remark For an example describing how to pass user data to a callback, see the second example
     *         for {@link WinccoaDpQueryConnectCallback}.
     * @see - {@link dpQuery}
     * - {@link dpQueryConnectSingle}
     * - {@link dpQueryDisconnect}
     * - {@link WinccoaConnectUpdateType}
     * - [CTRL function `dpQueryConnectAll()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpQueryConnectAll.html)
     * @example
     * ```ts
     * import {
     *   WinccoaManager,
     *   WinccoaConnectUpdateType,
     *   WinccoaError
     * } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * function  queryConnectCB (
     *   values: unknown[][],
     *   type: WinccoaConnectUpdateType,
     *   error?: WinccoaError
     * ) {
     *   if (error) {
     *     console.log(error);
     *     return;
     *   }
     *
     *   if (type == WinccoaConnectUpdateType.Answer)
     *     console.warn('--- Initial update ---');
     *
     *   for (let i = 0; i < values.length; i++) {
     *     console.info(`DPE = '%s', value = %s`, ...values[i]);
     *   }
     * }
     *
     * function connect() {
     *   let id = -1;
     *   try {
     *     id = winccoa.dpQueryConnectAll(
     *       queryConnectCB,
     *       true,
     *       `SELECT '_online.._value' FROM '*' WHERE _DPT="ExampleDP_Float"`
     *     );
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     *  ```
     */
    dpQueryConnectAll(callback, answer, query) {
        return this.api.dpQueryConnectAll(callback, answer, query);
    }
    //------------------------------------------------------------------------
    /**
     * Disconnect from {@link dpQueryConnectSingle} (or {@link dpQueryConnectAll}) .
     *
     * @param id ID of the connection to close as returned by {@link dpQueryConnectAll} or {@link dpQueryConnectSingle}.
     * @throws {@link WinccoaError} when __id__ is not found or invalid.
     * @returns ID of the closed connection (>= 0).
     * @see - {@link dpQueryConnectAll}
     *  - {@link dpQueryConnectSingle}
     * @example
     * ```ts
     * import {
     *   WinccoaManager,
     *   WinccoaConnectUpdateType,
     *   WinccoaError
     * } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * function connect() {
     * let id = -1;
     *   try {
     *     id = winccoa.dpQueryConnectSingle(
     *       function (
     *         values: unknown[][],
     *         type: WinccoaConnectUpdateType,
     *         error?: WinccoaError
     *       ) {
     *         if (error) {
     *           console.log(error);
     *           return;
     *         }
     *
     *         console.info(values)
     *       },
     *       true,
     *       "SELECT '_online.._value' FROM '*' WHERE _DPT= \"ExampleDP_Float\""
     *       );
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     *
     * function disconnect() {
     *   try {
     *     winccoa.dpQueryDisconnect(id);
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     * ```
     */
    dpQueryDisconnect(id) {
        return this.api.dpDisconnect(id);
    }
    //------------------------------------------------------------------------
    /**
     * Sets the current user ID to the specified value for the current instance of `WinccoaManager`.
     * @param id User ID to set.
     * @param password Password to use to set user (not needed if manager started as root).
     * @returns Boolean `true` in case of a success, otherwise {@link WinccoaError} wil be thrown instead.
     * @throws {@link WinccoaError} when __id__ is not found or invalid.
     * @see - [CTRL function `setUserId()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlS_Z/setUserId.html)
     * @example
     * ```ts
     * let isSuccess = false;
     * try {
     *   isSuccess = winccoa.setUserId(2048);  // operatorAll user id
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info('User id is set to operatorAll user id successfully - ' + isSuccess);
     * ```
     */
    setUserId(id, password) {
        return this.api.setUserId(id, password);
    }
    //------------------------------------------------------------------------
    /**
     * Reports a security event, this must be called whenever a security-relevant
     * action is made in JavaScript (like opening a server port).
     *
     * @param id ID of the security event.
     * @param args Arguments for the security event, depending on the value
     *             of id. See {@link WinccoaSecurityEventId} for details.
     * @throws {@link WinccoaError} when id is not known, required arguments
     *         are missing in args.
     * @see {@link WinccoaSecurityEventId}
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaSecurityEventId } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * winccoa.securityEvent(WinccoaSecurityEventId.PortOpened, 8443, 'https://');
     * ```
     */
    securityEvent(id, ...args) {
        return this.api.securityEvent(id, ...args);
    }
    //------------------------------------------------------------------------
    /**
     * This function returns the unit(s) of a data point.
     * @param dpeName Data point element
     * @returns Returns the unit as langText in one or several languages. In the event of an
     *          error, an empty string is returned.
     * > The returned data type can be defined with {@link setOptions}.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type.
     * @see - {@link dpSetUnit}
     * - {@link setOptions}
     * - [CTRL function `dpGetUnit()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpGetUnit.html)
     * @example
     * ```ts
     * try {
     *   let units = winccoa.dpGetUnit('ExampleDP_Rpt1.');
     *   console.info('DP units: ' + units);
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    dpGetUnit(dpeName) {
        return this.api.dpGetUnit(dpeName);
    }
    //------------------------------------------------------------------------------
    /**
     *  Sets the unit(s) for a data point.
     * @param dpeName Data point
     * @param unit Unit (for example, kg) in one or several languages as langText.
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type, etc.
     * @see - {@link dpGetUnit}
     * - [CTRL function `dpSetUnit()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetUnit.html)
     * @example
     * ```ts
     * let isSuccess = false;
     * try {
     *   isSuccess = await winccoa.dpSetUnit('ExampleDP_Rpt1.', {
          'de_AT.utf8': 's',
          'en_US.utf8': 's',
          'ru_RU.utf8': 'c',
        });
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     *   console.info('DP units are set successfully - ' + isSuccess);
     * ```
     */
    dpSetUnit(dpeName, unit) {
        return this.api.dpSetUnit(dpeName, unit);
    }
    //------------------------------------------------------------------------------
    /**
     * This function returns the numerical format(s) of a data point.
     * @param dpeName Data point element
     * @returns Returns a string that contains the format (for example, '%6.2f') in one or several
     *          languages or an empty string if an error occured.
     * > The returned data type can be defined with {@link setOptions}.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type.
     * @see - {@link dpSetFormat}
     * - [CTRL function `dpGetFormat()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpGetFormat.html)
     * @example
     * ```ts
     * let formats;
     * try {
     *   formats = winccoa.dpGetFormat('ExampleDP_Rpt1.');
     * } catch (exc) {
     *   console.error(exc);
     *   formats = null;
     * }
     *
     * if (formats)
     *   console.info('DP formats: ' + formats);
     * ```
     */
    dpGetFormat(dpeName) {
        return this.api.dpGetFormat(dpeName);
    }
    //------------------------------------------------------------------------------
    /**
     * Sets the numerical format(s) of a data point.
     * @param dpeName Data point element
     * @param format A string that contains the format (for example, '%6.2f') in one or several
     *                languages.
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type, etc.
     * @see - {@link dpGetFormat}
     * - [CTRL function `dpSetFormat()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetFormat.html)
     * @example
     * ```ts
     * let isSuccess = false;
     * try {
     *   isSuccess = await winccoa.dpSetFormat('ExampleDP_Rpt1.', {
          'de_AT.utf8': '%.4f',
          'en_US.utf8': '%.4f',
          'ru_RU.utf8': '%.2f',
        });
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     *   console.info('DP formats are set successfully - ' + isSuccess);
     * ```
     */
    dpSetFormat(dpeName, format) {
        return this.api.dpSetFormat(dpeName, format);
    }
    //------------------------------------------------------------------------------
    /**
     * Returns the comment (description) for the data point.
     * @param dpeName Data point element or data point
     * @param mode Mode of functionality.
     * For more details on all modes description see
     * [CTRL function `dpGetDescription()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpGetDescription.html).
     * @returns Description in all available languages as langText or empty strings in all
     *          languages.
     * > The returned data type can be defined with {@link setOptions}.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type.
     * @see - {@link dpSetDescription}
     * - [CTRL function `dpGetDescription()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpGetDescription.html)
     * @example
     * ```ts
     * let description;
     * try {
     *   description = winccoa.dpGetDescription('ExampleDP_Rpt1.');
     * } catch (exc) {
     *   console.error(exc);
     *   description = null;
     * }
     *
     * if (description)
     *   console.info('DP description: ' + description);
     * ```
     */
    dpGetDescription(dpeName, mode) {
        return this.api.dpGetDescription(dpeName, mode);
    }
    //------------------------------------------------------------------------------
    /**
     * Sets a comment (description) for the data point.
     * @param dpeName Data point element to be commented on
     * @param comment Comment as langText
     * @returns Promise - will be resolved to `true` if successful or rejected with an error.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type, etc.
     * @see - {@link dpGetDescription}
     * - [CTRL function `dpSetDescription()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetDescription.html)
     * @example
     * ```ts
     * let isSuccess = false;
     * try {
     *   isSuccess = await winccoa.dpSetDescription('ExampleDP_Rpt1.', {
          'de_AT.utf8': 'German description',
          'en_US.utf8': 'English description',
          'ru_RU.utf8': 'Russian description',
        });
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info('DP description is set successfully - ' + isSuccess);
     * ```
     */
    dpSetDescription(dpeName, comment) {
        return this.api.dpSetDescription(dpeName, comment);
    }
    //------------------------------------------------------------------------------
    /**
     * Returns the alias for the specified data point.
     * @param dpeName Data point element
     * @returns the appropriate alias in the language. Note that the alias can be only unilingual.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type.
     * @see - {@link dpSetAlias}
     * - [CTRL function `dpGetAlias()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpGetAlias.html)
     * @example
     * ```ts
     * try {
     *   let alias = winccoa.dpGetAlias('ExampleDP_Rpt1.');
     *   console.info('DP alias: ' + alias);
     * } catch (exc) {
     *   console.error(exc);
     * }
     * ```
     */
    dpGetAlias(dpeName) {
        return this.api.dpGetAlias(dpeName);
    }
    //------------------------------------------------------------------------------
    /**
     * Returns the data typeof a data point element.
     * @param dpeName Name of the data point element
     * @returns Type of a data point element
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type.
     * @see - {@link WinccoaElementType}
     * - [CTRL function `dpElementType()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpElementType.html)
     * @example
     * ```ts
     * import { WinccoaManager, WinccoaElementType } from 'winccoa-manager';
     * const winccoa = new WinccoaManager();
     *
     * async function dpElementTypeTest()
     *   try {
     *     let dpType = await winccoa.dpElementType('ExampleDP_Arg1.');
     *     console.info('The type of ExampleDP_Arg1 is float - ' + (dpType == WinccoaElementType.Float));
     *   } catch (exc) {
     *     console.error(exc);
     *   }
     * }
     * ```
     */
    dpElementType(dpeName) {
        return this.api.dpElementType(dpeName);
    }
    //------------------------------------------------------------------------------
    /**
     * Sets the alias for the specified data point element.
     * @param dpeName Data point element
     * @param alias Alias to be set. Note that the alias can be set only unilingual.
     * @returns Promise - will be resolved to true if successful or rejected with an error.
     * @throws {@link WinccoaError} when data point with the given __dpeName__ is not found or invalid argument type, etc.
     * @see - {@link dpGetAlias}
     * - [CTRL function `dpSetAlias()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlA_D/dpSetAlias.html)
     * @example
     * ```ts
     * let isSuccess = false;
     * try {
     *   isSuccess = await winccoa.dpSetAlias('ExampleDP_Rpt1.', 'rpt1Alias');
     * } catch (exc) {
     *   console.error(exc);
     * }
     *
     * console.info('DP alias is set successfully - ' + isSuccess);
     * ```
     */
    dpSetAlias(dpeName, alias) {
        return this.api.dpSetAlias(dpeName, alias);
    }
    //------------------------------------------------------------------------------
    /**
     * Returns the host name of the Event Manager this manager is connected to. Use this method only on redundant computers.
     * @returns Host name of the Event Manager this manager is connected to.
     * @see - [CTRL function `myReduHost()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlE_R/myReduHost.html)
     * @example
     * ```ts
     * const reduHostName = winccoa.myReduHost();
     * console.info('My redu host name - ' + reduHostName);
     * ```
     */
    myReduHost() {
        return this.api.myReduHost();
    }
    //------------------------------------------------------------------------
    /**
     * Returns host number in a redundant system depending on the connection to the Event Manager - manager 1 or 2 (for example, eventHost = "host1$host2").
     * In a non-redundant configuration this always returns 1.
     * @returns Host number.
     * @see - [CTRL function `myReduHostNum()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlE_R/myReduHostNum.html)
     * @example
     * ```ts
     * const reduHostNumber = winccoa.myReduHostNum();
     * console.info('My redu host number - ' + reduHostNumber);
     * ```
     */
    myReduHostNum() {
        return this.api.myReduHostNum();
    }
    //------------------------------------------------------------------------
    /**
     * Checks whether the project has been configured as redundant.
     * @returns `true` if project has been configured as redundant.
     * @see - [CTRL function `isRedundant()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlE_R/isRedundant.html)
     * @example
     * ```ts
     * const isProjectRedundant = winccoa.isRedundant();
     * console.info('My project is redundant - ' + isProjectRedundant);
     * ```
     */
    isRedundant() {
        return this.api.isRedundant();
    }
    //------------------------------------------------------------------------
    /**
     * Checks if the event manager to which this manager is connected is currently
     * the active REDU partner.
     * @returns `true` if the event manager to which this manager is connected is
     *          currently the active REDU partner.
     * @see - [CTRL function `isReduActive()`](https://www.winccoa.com/documentation/WinCCOA/latest/en_US/ControlE_R/isReduActive.html)
     * @example
     * ```ts
     * const isEventReduActive = winccoa.isReduActive();
     * console.info('Event currently is active - ' + isEventReduActive);
     * ```
     */
    isReduActive() {
        return this.api.isReduActive();
    }
}
exports.WinccoaManager = WinccoaManager;
