// winccoa-manager/lib/ctrl-types.js
"use strict";
// NB: Definitions in here need to kept in sync with C++ code, in particular:
// - enum VariableType is a copy of the C++ enum VariableType (from Variable.hxx),
//        but now in TypeScript.
// - enum WinccoaCtrlType maps a subset of the type names as used in CTRL code to
//        values in VariableType
Object.defineProperty(exports, "__esModule", { value: true });
exports.WinccoaCtrlType = void 0;
/*
 * Enumerates variable types that can be used as parameters for CTRL function calls
 * using the same names as C++ enum VariableType
 */
var VariableType;
(function (VariableType) {
    /** (0x00000000). unset variable type */
    VariableType[VariableType["NO_VAR"] = 0] = "NO_VAR";
    /** (0x00010000). void variable type */
    VariableType[VariableType["NOTYPE_VAR"] = 65536] = "NOTYPE_VAR";
    /** (0x00020000). empty variable type */
    VariableType[VariableType["VARIABLE"] = 131072] = "VARIABLE";
    /** (0x00030000). time variable type */
    VariableType[VariableType["TIME_VAR"] = 196608] = "TIME_VAR";
    /** (0x00040000). bit variable type */
    VariableType[VariableType["BIT_VAR"] = 262144] = "BIT_VAR";
    /** (0x00050000). integer variable type */
    VariableType[VariableType["INTEGER_VAR"] = 327680] = "INTEGER_VAR";
    /** (0x00060000). unsigned integer variable type */
    VariableType[VariableType["UINTEGER_VAR"] = 393216] = "UINTEGER_VAR";
    /** (0x00070000). float variable type */
    VariableType[VariableType["FLOAT_VAR"] = 458752] = "FLOAT_VAR";
    /** (0x00080000). text variable type */
    VariableType[VariableType["TEXT_VAR"] = 524288] = "TEXT_VAR";
    /** (0x00090000). 32 bit variable type */
    VariableType[VariableType["BIT32_VAR"] = 589824] = "BIT32_VAR";
    /** (0x000A0000). char variable type */
    VariableType[VariableType["CHAR_VAR"] = 655360] = "CHAR_VAR";
    /** (0x000B0000). dp identifier variable type */
    VariableType[VariableType["DPIDENTIFIER_VAR"] = 720896] = "DPIDENTIFIER_VAR";
    /** (0x000C0000). empty dynvar type */
    VariableType[VariableType["DYN_VAR"] = 786432] = "DYN_VAR";
    /** (0x000D0000). dyn_time variable type */
    VariableType[VariableType["DYNTIME_VAR"] = 851968] = "DYNTIME_VAR";
    /** (0x000E0000). dyn_bit variable type */
    VariableType[VariableType["DYNBIT_VAR"] = 917504] = "DYNBIT_VAR";
    /** (0x000F0000). dyn_int variable type */
    VariableType[VariableType["DYNINTEGER_VAR"] = 983040] = "DYNINTEGER_VAR";
    /** (0x00100000). dyn_uint variable type */
    VariableType[VariableType["DYNUINTEGER_VAR"] = 1048576] = "DYNUINTEGER_VAR";
    /** (0x00110000). dyn_float variable type */
    VariableType[VariableType["DYNFLOAT_VAR"] = 1114112] = "DYNFLOAT_VAR";
    /** (0x00120000). dyn_text variable type */
    VariableType[VariableType["DYNTEXT_VAR"] = 1179648] = "DYNTEXT_VAR";
    /** (0x00130000). dyn_32bit variable type */
    VariableType[VariableType["DYNBIT32_VAR"] = 1245184] = "DYNBIT32_VAR";
    /** (0x00140000). dyn_char variable type */
    VariableType[VariableType["DYNCHAR_VAR"] = 1310720] = "DYNCHAR_VAR";
    /** (0x00150000). dyn_dpidentifier variable type */
    VariableType[VariableType["DYNDPIDENTIFIER_VAR"] = 1376256] = "DYNDPIDENTIFIER_VAR";
    /** (0x00160000). recvar  type */
    VariableType[VariableType["REC_VAR"] = 1441792] = "REC_VAR";
    /** (0x00170000). file variable type */
    VariableType[VariableType["FILE_VAR"] = 1507328] = "FILE_VAR";
    /** (0x00180000). empty anytype variable type */
    VariableType[VariableType["ANYTYPE_VAR"] = 1572864] = "ANYTYPE_VAR";
    /** (0x00190000). argument variable type (internal use)*/
    VariableType[VariableType["ARG_VAR"] = 1638400] = "ARG_VAR";
    /** (0x001A0000). dyn_dyn_time variable type*/
    VariableType[VariableType["DYNDYNTIME_VAR"] = 1703936] = "DYNDYNTIME_VAR";
    /** (0x001B0000). dyn_dyn_bit variable type*/
    VariableType[VariableType["DYNDYNBIT_VAR"] = 1769472] = "DYNDYNBIT_VAR";
    /** (0x001C0000). dyn_dyn_int variable type*/
    VariableType[VariableType["DYNDYNINTEGER_VAR"] = 1835008] = "DYNDYNINTEGER_VAR";
    /** (0x001D0000). dyn_dyn_uint variable type*/
    VariableType[VariableType["DYNDYNUINTEGER_VAR"] = 1900544] = "DYNDYNUINTEGER_VAR";
    /** (0x001E0000). dyn_dyn_float variable type*/
    VariableType[VariableType["DYNDYNFLOAT_VAR"] = 1966080] = "DYNDYNFLOAT_VAR";
    /** (0x001F0000). dyn_dyn_text variable type*/
    VariableType[VariableType["DYNDYNTEXT_VAR"] = 2031616] = "DYNDYNTEXT_VAR";
    /** (0x00200000). dyn_dyn_32bit variable type*/
    VariableType[VariableType["DYNDYNBIT32_VAR"] = 2097152] = "DYNDYNBIT32_VAR";
    /** (0x00210000). dyn_dyn_char variable type*/
    VariableType[VariableType["DYNDYNCHAR_VAR"] = 2162688] = "DYNDYNCHAR_VAR";
    /** (0x00220000). dyn_dyn_dpidentifier variable type*/
    VariableType[VariableType["DYNDYNDPIDENTIFIER_VAR"] = 2228224] = "DYNDYNDPIDENTIFIER_VAR";
    /** (0x00230000). dyn_anytype variable type*/
    VariableType[VariableType["DYNANYTYPE_VAR"] = 2293760] = "DYNANYTYPE_VAR";
    /** (0x00240000). dyn_dyn_anytype variable type*/
    VariableType[VariableType["DYNDYNANYTYPE_VAR"] = 2359296] = "DYNDYNANYTYPE_VAR";
    /** (0x00250000). extended_time variable type*/
    VariableType[VariableType["EXTTIME_VAR"] = 2424832] = "EXTTIME_VAR";
    /** (0x00260000). dyn_extended_time variable type*/
    VariableType[VariableType["DYNEXTTIME_VAR"] = 2490368] = "DYNEXTTIME_VAR";
    /** (0x00270000). dyn_dyn_extended_time variable type*/
    VariableType[VariableType["DYNDYNEXTTIME_VAR"] = 2555904] = "DYNDYNEXTTIME_VAR";
    /** (0x00280000). langtext variable type*/
    VariableType[VariableType["LANGTEXT_VAR"] = 2621440] = "LANGTEXT_VAR";
    /** (0x00290000). dyn_langtext variable type*/
    VariableType[VariableType["DYNLANGTEXT_VAR"] = 2686976] = "DYNLANGTEXT_VAR";
    /** (0x002A0000). dyn_dyn_langtext variable type*/
    VariableType[VariableType["DYNDYNLANGTEXT_VAR"] = 2752512] = "DYNDYNLANGTEXT_VAR";
    /** (0x002B0000). error variable type*/
    VariableType[VariableType["ERROR_VAR"] = 2818048] = "ERROR_VAR";
    /** (0x002C0000). dyn_error variable type*/
    VariableType[VariableType["DYNERROR_VAR"] = 2883584] = "DYNERROR_VAR";
    /** (0x002D0000). dyn_dyn_error variable type*/
    VariableType[VariableType["DYNDYNERROR_VAR"] = 2949120] = "DYNDYNERROR_VAR";
    /** (0x002E0000). blob variable type */
    VariableType[VariableType["BLOB_VAR"] = 3014656] = "BLOB_VAR";
    /** (0x002F0000). dyn_blob variable type */
    VariableType[VariableType["DYNBLOB_VAR"] = 3080192] = "DYNBLOB_VAR";
    /** (0x00310000). ADO data types          */
    VariableType[VariableType["RECHDL_VAR"] = 3211264] = "RECHDL_VAR";
    VariableType[VariableType["CONNHDL_VAR"] = 3276800] = "CONNHDL_VAR";
    VariableType[VariableType["CMDHDL_VAR"] = 3342336] = "CMDHDL_VAR";
    VariableType[VariableType["POINTER_VAR"] = 3407872] = "POINTER_VAR"; /* in Libs/Ctrl */
    VariableType[VariableType["SHAPE_VAR"] = 3473408] = "SHAPE_VAR"; /* shape type - UI */
    VariableType[VariableType["IDISPATCH_VAR"] = 3538944] = "IDISPATCH_VAR"; /* IDispatch type - UI */
    VariableType[VariableType["DYNDYNBLOB_VAR"] = 3604480] = "DYNDYNBLOB_VAR";
    VariableType[VariableType["DYNREC_VAR"] = 3670016] = "DYNREC_VAR";
    VariableType[VariableType["DYNDYNREC_VAR"] = 3735552] = "DYNDYNREC_VAR";
    VariableType[VariableType["MAPPING_VAR"] = 3801088] = "MAPPING_VAR";
    VariableType[VariableType["DYNMAPPING_VAR"] = 3866624] = "DYNMAPPING_VAR";
    VariableType[VariableType["DYNDYNMAPPING_VAR"] = 3932160] = "DYNDYNMAPPING_VAR";
    VariableType[VariableType["MIXED_VAR"] = 3997696] = "MIXED_VAR";
    VariableType[VariableType["DYNMIXED_VAR"] = 4063232] = "DYNMIXED_VAR";
    VariableType[VariableType["DYNDYNMIXED_VAR"] = 4128768] = "DYNDYNMIXED_VAR";
    VariableType[VariableType["DYNRECHDL_VAR"] = 4194304] = "DYNRECHDL_VAR";
    VariableType[VariableType["DYNCONNHDL_VAR"] = 4259840] = "DYNCONNHDL_VAR";
    VariableType[VariableType["DYNCMDHDL_VAR"] = 4325376] = "DYNCMDHDL_VAR";
    VariableType[VariableType["DYNDYNRECHDL_VAR"] = 4390912] = "DYNDYNRECHDL_VAR";
    VariableType[VariableType["DYNDYNCONNHDL_VAR"] = 4456448] = "DYNDYNCONNHDL_VAR";
    VariableType[VariableType["DYNDYNCMDHDL_VAR"] = 4521984] = "DYNDYNCMDHDL_VAR";
    VariableType[VariableType["LONG_VAR"] = 4587520] = "LONG_VAR";
    VariableType[VariableType["DYNLONG_VAR"] = 4653056] = "DYNLONG_VAR";
    VariableType[VariableType["DYNDYNLONG_VAR"] = 4718592] = "DYNDYNLONG_VAR";
    VariableType[VariableType["ULONG_VAR"] = 4784128] = "ULONG_VAR";
    VariableType[VariableType["DYNULONG_VAR"] = 4849664] = "DYNULONG_VAR";
    VariableType[VariableType["DYNDYNULONG_VAR"] = 4915200] = "DYNDYNULONG_VAR";
    /*IM 87586: new _original, _online, _offline attribute _status64 */
    /** (0x004C0000). 64 bit variable type */
    VariableType[VariableType["BIT64_VAR"] = 4980736] = "BIT64_VAR";
    VariableType[VariableType["DYNBIT64_VAR"] = 5046272] = "DYNBIT64_VAR";
    VariableType[VariableType["DYNDYNBIT64_VAR"] = 5111808] = "DYNDYNBIT64_VAR";
    VariableType[VariableType["DYNSHAPE_VAR"] = 5177344] = "DYNSHAPE_VAR";
    VariableType[VariableType["DYNDYNSHAPE_VAR"] = 5242880] = "DYNDYNSHAPE_VAR";
    // defined in Libs/Ctrl
    VariableType[VariableType["FUNCTION_VAR"] = 5308416] = "FUNCTION_VAR";
    VariableType[VariableType["DYNFUNCTION_VAR"] = 5373952] = "DYNFUNCTION_VAR";
    VariableType[VariableType["DYNDYNFUNCTION_VAR"] = 5439488] = "DYNDYNFUNCTION_VAR";
    VariableType[VariableType["REFERENCE_VAR"] = 5505024] = "REFERENCE_VAR";
    VariableType[VariableType["CLASS_VAR"] = 5570560] = "CLASS_VAR";
    VariableType[VariableType["VECTOR_VAR"] = 5636096] = "VECTOR_VAR";
    VariableType[VariableType["VOLATILE_VAR"] = 5701632] = "VOLATILE_VAR";
    // QVariant in Qt-Ui
    VariableType[VariableType["QVARIANT_VAR"] = 5767168] = "QVARIANT_VAR";
})(VariableType || (VariableType = {}));
/**
 * Enumerates CTRL data types that can be used as parameters for CTRL function calls.
 * The enumerated values have exactly the same name as the corresponding CTRL type
 * and therefore start with a lowercase letter.
 * @see {@link WinccoaCtrlScript.start}
 */
var WinccoaCtrlType;
(function (WinccoaCtrlType) {
    /** `time` variable type */
    WinccoaCtrlType[WinccoaCtrlType["time"] = 196608] = "time";
    /** `bool` variable type */
    WinccoaCtrlType[WinccoaCtrlType["bool"] = 262144] = "bool";
    /** `int` variable type */
    WinccoaCtrlType[WinccoaCtrlType["int"] = 327680] = "int";
    /** `uint` variable type */
    WinccoaCtrlType[WinccoaCtrlType["uint"] = 393216] = "uint";
    /** `float` variable type */
    WinccoaCtrlType[WinccoaCtrlType["float"] = 458752] = "float";
    /** `double` variable type */
    WinccoaCtrlType[WinccoaCtrlType["double"] = 458752] = "double";
    /** `string` variable type */
    WinccoaCtrlType[WinccoaCtrlType["string"] = 524288] = "string";
    /** `bit32` variable type */
    WinccoaCtrlType[WinccoaCtrlType["bit32"] = 589824] = "bit32";
    /** `char` variable type */
    WinccoaCtrlType[WinccoaCtrlType["char"] = 655360] = "char";
    /** `dyn_time` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_time"] = 851968] = "dyn_time";
    /** `dyn_bool` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_bool"] = 917504] = "dyn_bool";
    /** `dyn_int` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_int"] = 983040] = "dyn_int";
    /** `dyn_uint` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_uint"] = 1048576] = "dyn_uint";
    /** `dyn_float` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_float"] = 1114112] = "dyn_float";
    /** `dyn_string` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_string"] = 1179648] = "dyn_string";
    /** `dyn_32bit` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_bit32"] = 1245184] = "dyn_bit32";
    /** `dyn_char` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_char"] = 1310720] = "dyn_char";
    /** `dyn_dyn_time` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_time"] = 1703936] = "dyn_dyn_time";
    /** `dyn_dyn_bool` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_bool"] = 1769472] = "dyn_dyn_bool";
    /** `dyn_dyn_int` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_int"] = 1835008] = "dyn_dyn_int";
    /** `dyn_dyn_uint` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_uint"] = 1900544] = "dyn_dyn_uint";
    /** `dyn_dyn_float` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_float"] = 1966080] = "dyn_dyn_float";
    /** `dyn_dyn_string` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_string"] = 2031616] = "dyn_dyn_string";
    /** `dyn_dyn_32bit` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_bit32"] = 2097152] = "dyn_dyn_bit32";
    /** `dyn_dyn_char` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_char"] = 2162688] = "dyn_dyn_char";
    /** `atime` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["atime"] = 2424832] = "atime";
    /** `dyn_atime` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_atime"] = 2490368] = "dyn_atime";
    /** `dyn_dyn_atime` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_atime"] = 2555904] = "dyn_dyn_atime";
    /** `langString` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["langString"] = 2621440] = "langString";
    /** `dyn_langString` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_langString"] = 2686976] = "dyn_langString";
    /** `dyn_dyn_langString variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_langString"] = 2752512] = "dyn_dyn_langString";
    /** `long` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["long"] = 4587520] = "long";
    /** `dyn_long` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_long"] = 4653056] = "dyn_long";
    /** `dyn_dyn_long` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_long"] = 4718592] = "dyn_dyn_long";
    /** `ulong` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["ulong"] = 4784128] = "ulong";
    /** `dyn_ulong` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_ulong"] = 4849664] = "dyn_ulong";
    /** `dyn_dyn_ulong` variable type*/
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_ulong"] = 4915200] = "dyn_dyn_ulong";
    /** `bit64` variable type */
    WinccoaCtrlType[WinccoaCtrlType["bit64"] = 4980736] = "bit64";
    /** `dyn_bit6` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_bit64"] = 5046272] = "dyn_bit64";
    /** `dyn_dyn_bit64` variable type */
    WinccoaCtrlType[WinccoaCtrlType["dyn_dyn_bit64"] = 5111808] = "dyn_dyn_bit64";
})(WinccoaCtrlType || (exports.WinccoaCtrlType = WinccoaCtrlType = {}));
