// winccoa-manager/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.delay = exports.WinccoaCtrlType = exports.WinccoaCtrlScript = exports.WinccoaLangTextFormat = exports.WinccoaAlertTime = exports.WinccoaError = exports.WinccoaSecurityEventId = exports.WinccoaElementType = exports.WinccoaConnectUpdateType = exports.WinccoaManager = void 0;
var binding_1 = require("./lib/binding");
Object.defineProperty(exports, "WinccoaManager", { enumerable: true, get: function () { return binding_1.WinccoaManager; } });
Object.defineProperty(exports, "WinccoaConnectUpdateType", { enumerable: true, get: function () { return binding_1.WinccoaConnectUpdateType; } });
Object.defineProperty(exports, "WinccoaElementType", { enumerable: true, get: function () { return binding_1.WinccoaElementType; } });
Object.defineProperty(exports, "WinccoaSecurityEventId", { enumerable: true, get: function () { return binding_1.WinccoaSecurityEventId; } });
var error_1 = require("./lib/error");
Object.defineProperty(exports, "WinccoaError", { enumerable: true, get: function () { return error_1.WinccoaError; } });
var alerttime_1 = require("./lib/alerttime");
Object.defineProperty(exports, "WinccoaAlertTime", { enumerable: true, get: function () { return alerttime_1.WinccoaAlertTime; } });
var options_1 = require("./lib/options");
Object.defineProperty(exports, "WinccoaLangTextFormat", { enumerable: true, get: function () { return options_1.WinccoaLangTextFormat; } });
var ctrl_script_1 = require("./lib/ctrl-script");
Object.defineProperty(exports, "WinccoaCtrlScript", { enumerable: true, get: function () { return ctrl_script_1.WinccoaCtrlScript; } });
var ctrl_types_1 = require("./lib/ctrl-types");
Object.defineProperty(exports, "WinccoaCtrlType", { enumerable: true, get: function () { return ctrl_types_1.WinccoaCtrlType; } });
var tools_1 = require("./lib/tools");
Object.defineProperty(exports, "delay", { enumerable: true, get: function () { return tools_1.delay; } });
